### DISCUSSION TITLE
Maintaining WiFi connection on Raspberry Pi

### DISCUSSION THREAD

**Discussion by dalogue1** [Nachtzuster/BirdNET-Pi#218] [General]
I have had issues with my four Raspberry Pi BirdNET-Pis becoming disconnecting from the WiFi. (Multiple Pis, different locations, different Access Points) It turns out to not be an uncommon problem.  The best solution to date is to configure a watchdog to reestablish the wlan0 connection and to disable the power saving mode for the WiFi. This is applicable regardless of Raspberry Pi or whether bookworm or bullseye.

**Watchdog Configuration**
The following web link describes the monitor process:  
https://www.dzombak.com/blog/2023/12/Maintaining-a-solid-WiFi-connection-on-Raspberry-Pi.html

**Turning Off the Power Saving Mode**
The problem appears to be related to a power saving mode option that can be turned off by creating a configuration file named **/etc/modprobe.d/rtl8192cu.conf**  containing the following and then rebooting:

**options rtl8192cu rtw_power_mgnt=0 rtw_enusbss=0 rtw_ips_mode=1**

_See the last message in this link https://forums.raspberrypi.com/viewtopic.php?t=278409
However, note that the name of the driver has since 2020 been changed to rtl8192cu instead of 8192cu as stated in the link._

**Comment by alexbelgium** [comment]
Thanks for sharing!

**Reply by dalogue1** [reply]
BTW Turning Off the Power Saving Mode was enough to eliminate the dropping connection.  The Watchdog Configuration is a backup measure.

**Comment by dalogue1** [comment]
You can create the configuration file by entering the following commands using the Web Terminal:

_sudo -s
cat > /etc/modprobe.d/rtl8192cu.conf
options rtl8192cu rtw_power_mgnt=0 rtw_enusbss=0 rtw_ips_mode=1
^D
exit_

After creating the file as shown above, simply reboot and the Power Saving Mode is disabled.
_(Note: the ^D implies hitting "Control D" not typing the string.)_

**Comment by dalogue1** [comment]
This link is to a permanent way to disable power saving on WiFi.

https://unix.stackexchange.com/questions/269661/how-to-turn-off-wireless-power-management-permanently

The link describes in detail a complicated procedure which is good for understanding the fix, but basically you can do the following and have the same effect:

Run the following command: 
**sudo crontab -e**

Use the editor to put the following line at the end of the file:
**\* * * * * sudo /sbin/iwconfig wlan0 power off**

Exit the editor and reboot.

Note: You no longer need any reference to power management in the **/etc/rc.local** file if one was added.
